retrieval/discovery/serverset.go

```diff
@@ -0,0 +1,359 @@
+	conn.SetLogger(ZookeeperLogger{})
+	if err != nil {
+		return nil
+	}
+	updates := make(chan ZookeeperTreeCacheEvent, 20000)
+	sd := &ServersetDiscovery{
+		conf:      conf,
+		conn:      conn,
+		updates:   updates,
+		sources:   make(map[string]struct{}),
+		runDone:   make(chan struct{}),
+		treeCache: NewZookeeperTreeCache(conn, conf.Paths[0], updates),
+	}
+	return sd
+}
+
+// Sources implements the TargetProvider interface.
+func (sd *ServersetDiscovery) Sources() []string {
+	sd.mu.RLock()
+	defer sd.mu.RUnlock()
+	srcs := []string{}
+	for t := range sd.sources {
+		srcs = append(srcs, t)
+	}
+	return srcs
+}
+
+// Run implements the TargetProvider interface.
+func (sd *ServersetDiscovery) Run(ch chan<- *config.TargetGroup) {
+	defer close(ch)
+	defer sd.conn.Close()
+
+	for {
+		select {
+		case event := <-sd.updates:
+			tg := &config.TargetGroup{
+				Source: serversetSourcePrefix + event.Path,
+			}
+			sd.mu.Lock()
+			if event.Data != nil {
+				labelSet, err := parseServersetMember(*event.Data, event.Path)
+				if err != nil {
+					sd.sources[event.Path] = struct{}{}
+					sd.mu.Unlock()
+					continue
+				} else {
+					delete(sd.sources, event.Path)
+				}
+				tg.Targets = []clientmodel.LabelSet{*labelSet}
+			} else {
+				delete(sd.sources, event.Path)
+			}
+			sd.mu.Unlock()
+			ch <- tg
+		case <-sd.runDone:
+			sd.treeCache.Stop()
+			return
+		}
+	}
+}
+
+// Stop implements the TargetProvider interface.
+func (sd *ServersetDiscovery) Stop() {
+	log.Debugf("Stopping serverset service discovery for %s %s", sd.conf.Servers, sd.conf.Paths)
+
+	// Terminate Run.
+	sd.runDone <- struct{}{}
+
+	log.Debugf("Serverset service discovery for %s %s stopped", sd.conf.Servers, sd.conf.Paths)
+}
+
+func parseServersetMember(data []byte, path string) (*clientmodel.LabelSet, error) {
+	member := serversetMember{}
+	err := json.Unmarshal(data, &member)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshaling serverset member %q: %s", path, err)
+	}
+
+	labels := clientmodel.LabelSet{}
+	labels[serversetPathLabel] = clientmodel.LabelValue(path)
+	labels[clientmodel.AddressLabel] = clientmodel.LabelValue(
+		fmt.Sprintf("%s:%d", member.ServiceEndpoint.Host, member.ServiceEndpoint.Port))
+
+	labels[serversetEndpointLabelPrefix+"_host"] = clientmodel.LabelValue(member.ServiceEndpoint.Host)
+	labels[serversetEndpointLabelPrefix+"_port"] = clientmodel.LabelValue(fmt.Sprintf("%d", member.ServiceEndpoint.Port))
+
+	for name, endpoint := range member.AdditionalEndpoints {
+		cleanName := clientmodel.LabelName(invalidLabelCharRE.ReplaceAllString(name, "_"))
+		labels[serversetEndpointLabelPrefix+"_host_"+cleanName] = clientmodel.LabelValue(
+			endpoint.Host)
+		labels[serversetEndpointLabelPrefix+"_port_"+cleanName] = clientmodel.LabelValue(
+			fmt.Sprintf("%d", endpoint.Port))
+
+	}
+
+	labels[serversetStatusLabel] = clientmodel.LabelValue(member.Status)
+
+	return &labels, nil
+}
+
+type ZookeeperTreeCache struct {
```

Comment:
Yes, I'm planning on offering this upstream as an initial Go version of https://curator.apache.org/apidocs/org/apache/curator/framework/recipes/cache/TreeCache.html so I was setting visibility accordingly.